Add documentation on batch (#11)
SAP/cloud-sdk: docs/js/features/odata/common/batch/changeset.mdx

A changeset is a collection of HTTP `POST`, `PUT`, `PATCH` and `DELETE` operations - requests built by any [CreateRequestBuilder](#create-request-builder), [UpdateRequestBuilder](#update-request-builder) and [DeleteRequestBuilder](#delete-request-builder) in terms of the SAP Cloud SDK.
The order of execution within a changeset is not defined.
This differs from the whole batch request itself, where the order is defined.
Therefore the requests within a changeset should not depend on each other.
If the execution of any of the requests within a changeset fails, the whole changeset will be reflected as an error in the response and will not be applied, much like a database transaction.

> Change request cannot be passed to a batch request directly.

Reviewer: Either 
> change requests cannot ...
or
> A change request cannot ...

> Change request cannot be passed to a batch request directly.

Author: ```suggestion
Change requests cannot be passed to a batch request directly.
```

> They have to be combined in a changset, which in turn an be passed to the batch request.

Reviewer: ```suggestion
They have to be combined in a changset, which in turn can be passed to the batch request.
```

Once a batch request is executed it returns a list of `BatchResponse`s.

> Those contain the raw response information of each subrequest, the subresponse of a changeset request can either be `WriteResponses` or an `ErrorResponse`.

Reviewer: ```suggestion
Those contain the raw response information of each subrequest.
The subresponse of a changeset request can either be a `WriteResponse` or an `ErrorResponse`.
```

Reviewer: I changed WriteResponses to the singular form. Not sure if the plural was intended!

Author: No it was correct. I know it's weird 🙁

Reviewer: Maybe this can be improved by pointing out that there could be a collection of responses called `WriteResponses` or in case of failure a single `ErrorResponse`?

To determine if a request was successful use `.isSuccess()`.

> Successful requests can be cast to `WriteResponses` which contains all subresponses for the changeset request.

Reviewer: In the JS context, how do you cast something to a type? (in typescript it's `as NEW_TYPE` but you can always do that, it's just sometimes wrong).

Author: That is what is meant here, I will rephrase it to 'should'

> Those reponses can be accessed by `.responses` and have the type `WriteResponse`.

Author: ```suggestion
Those responses can be accessed by `.responses` and have the type `WriteResponse`.
```

> Each `WriteResponse` contains the HTTP code and can contain the raw body and the constructor of the entity that was parsed from the response, depending on whether there was a response.

Reviewer: > , depending on whether there was a response.

How can the batch succeed if there was no response at all for a part of a changeset? I guess what is meant is that there the response had no body.

Author: Correct, I adjusted it.

> Each `WriteResponse` contains the HTTP code and can contain the raw body and the constructor of the entity that was parsed from the response, depending on whether there was a response.

Author: ```suggestion
Each `WriteResponse` contains the HTTP code and can contain the raw body and the constructor of the entity that was parsed from the response, depending on whether there was a body in the response.
```

> Create requests typically do not return a response.

Reviewer: response or body (aka `204 No Content` or `201 Created` is still a response in my understanding)

Author: Yeah, I fixed it.

> Create requests typically do not return a response.

Author: ```suggestion
Create and delete requests typically do not have a response body.
```

> In order to work with an instance of an entity given in a `WriteResponse`, you can use the `.as` method, that allows you to transform the raw data into an instance of the given constructor.

Reviewer: Consider adding that the raw body is a string for people that may not be as experienced.

> In order to work with an instance of an entity given in a `WriteResponse`, you can use the `.as` method, that allows you to transform the raw data into an instance of the given constructor.

Author: ```suggestion
In order to work with an instance of an entity given in a `WriteResponse`, you can use the `.as` method, that allows you to transform the raw string body into an instance of the given constructor.
```

> Note that the response cannot exist, so you should only call this method if you know that there is data.

Reviewer: ```suggestion
Note that the response may not exist, so you should only call this method if you know that there is data.
```

Typically the HTTP code is a good indicator for this (`201 No Content` probably won't have content).
If you are working with TypeScript you will have to tell the compiler, that the `.as!` method can be used here by adding a `!`.
Also note, that retrieve responses can be [`ErrorResponse`s](#error-handling).

> Therefore, it is crutial to check responses for success, before casting them to `WriteResponses`.

Reviewer: ```suggestion
Therefore, it is crucial to check responses for success, before casting them to `WriteResponses`.
```

A changeset is a collection of HTTP `POST`, `PUT`, `PATCH` and `DELETE` operations - requests built by any [CreateRequestBuilder](#create-request-builder), [UpdateRequestBuilder](#update-request-builder) and [DeleteRequestBuilder](#delete-request-builder) in terms of the SAP Cloud SDK.
The order of execution within a changeset is not defined.
This differs from the whole batch request itself, where the order is defined.
Therefore the requests within a changeset should not depend on each other.
If the execution of any of the requests within a changeset fails, the whole changeset will be reflected as an error in the response and will not be applied, much like a database transaction.

Change requests cannot be passed to a batch request directly.
They have to be combined in a changset, which in turn can be passed to the batch request.
Once a batch request is executed it returns a list of `BatchResponse`s.
Those contain the raw response information of each subrequest.

> The subresponse of a changeset request can either be `WriteResponses` or an `ErrorResponse`.

Reviewer: Maybe this can be improved by pointing out that there could be a collection of responses called `WriteResponses` or in case of failure a single `ErrorResponse`?

> The subresponse of a changeset request can either be `WriteResponses` or an `ErrorResponse`.

Reviewer: Maybe this can be improved by pointing out that there could be a collection of responses called `WriteResponses` or in case of failure a single `ErrorResponse`? (sorry if it was posted multiple times... github trolled me a bit)

Author: I rephrased it a bit. I found it difficult to express because the name of the type is not so great... Maybe something for the 2.0 discussion...

> The subresponse of a changeset request can either be `WriteResponses` or an `ErrorResponse`.

Author: ```suggestion
The response to a changeset request can either be a collection of the subresponses to the subrequests of the changeset of type `WriteResponses` or an `ErrorResponse`.
```
